From patches Mon Sep 17 00:00:00 2001
From: Wenchao Xia <xiawenc@linux.vnet.ibm.com>
To: qemu-devel@nongnu.org
Cc: kwolf@redhat.com, pbonzini@redhat.com,
 Wenchao Xia <xiawenc@linux.vnet.ibm.com>, dietmar@proxmox.com,
 stefanha@gmail.com
Date: Sat, 13 Apr 2013 19:11:40 +0800
Message-id: 1365851501-3037-5-git-send-email-xiawenc@linux.vnet.ibm.com
Subject: [PATCH 4/5] block: package rolling back code in qmp_transaction()

Signed-off-by: Wenchao Xia <xiawenc@linux.vnet.ibm.com>
Reviewed-by: Eric Blake <eblake@redhat.com>
---
 blockdev.c |   12 +++++++++---
 1 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/blockdev.c b/blockdev.c
index 38cb7d5..3e69569 100644
--- a/blockdev.c
+++ b/blockdev.c
@@ -889,6 +889,14 @@ static void external_snapshot_commit(BlockdevAction *action,
                 NULL);
 }
 
+static void external_snapshot_rollback(BlockdevAction *action,
+                                       BlkTransactionStates *states)
+{
+    if (states->new_bs) {
+        bdrv_delete(states->new_bs);
+    }
+}
+
 /*
  * 'Atomic' group snapshots.  The snapshots are taken as a set, and if any fail
  *  then we do not pivot any of the devices in the group, and abandon the
@@ -943,9 +951,7 @@ delete_and_fail:
     * the original bs for all images
     */
     QSIMPLEQ_FOREACH(states, &snap_bdrv_states, entry) {
-        if (states->new_bs) {
-             bdrv_delete(states->new_bs);
-        }
+        external_snapshot_rollback(NULL, states);
     }
 exit:
     QSIMPLEQ_FOREACH_SAFE(states, &snap_bdrv_states, entry, next) {
-- 
1.7.1



